### DISCUSSION TITLE
Deforum extension mounts (auto)

### DISCUSSION THREAD

**Discussion by timelf123** [AbdBarho/stable-diffusion-webui-docker#255] [Ideas]
Not sure how you want to handle extension dependencies if at all... But people are using Deforum in auto's ui currently, the models need to be downloaded every time the container is rebuilt. 

What is missing from the Deforum extension experience:
- Add ffmpeg to the auto docker container (to generate videos inline)
- `MOUNTS["${ROOT}/models/Deforum"]="/data/Deforum"` to store the models used

EDIT:
I just realized I can solve missing dependecies with a script `data/config/auto/startup.sh`:
```
set -e
# Deforum
echo "Installing Deforum dependencies"
MOUNTS["${ROOT}/models/Deforum"]="/data/Deforum"
apt-get install wget ffmpeg -yq
echo "Done installing Deforum dependencies"
```

`entrypoint.sh` - switched custom script and mounting logic
```
# run custom script first before mounting
if [ -f "/data/config/auto/startup.sh" ]; then
  pushd ${ROOT}
  . /data/config/auto/startup.sh
  popd
fi

for to_path in "${!MOUNTS[@]}"; do
  set -Eeuo pipefail
  from_path="${MOUNTS[${to_path}]}"
  rm -rf "${to_path}"
  if [ ! -f "$from_path" ]; then
    mkdir -vp "$from_path"
  fi
  mkdir -vp "$(dirname "${to_path}")"
  ln -sT "${from_path}" "${to_path}"
  echo Mounted $(basename "${from_path}")
done
```

**Are there adverse side effects from making the custom script run before the mounter?** I'd guess you'd prefer to source custom mount paths from a `startup.sh` file rather than defining them in the entrypoint? In that case, switching the order of the mounting/startup script allows MOUNT points to be added since the startup.sh is sourced

**Comment by AbdBarho** [comment]
@timelf123  why not just mount directly in the script

```sh
set -e
# Deforum
echo "Installing Deforum dependencies"
apt-get install wget ffmpeg -yq
mkdir -vp /data/Deforum
ln -sT /data/Deforum "${ROOT}/models/Deforum" 
echo "Done installing Deforum dependencies"
```

what is the advantage of re-using the `MOUNTS` variable?

**Reply by timelf123** [reply]
🤦 I didn't even consider that as an option. Thanks!

**Comment by timelf123** [comment]
thanks again for the idea. For anyone else coming here later: I landed on this (added test to only create symlink if one doesn't exist):
```bash
set -e

echo "> Deforum dependencies"
# ffmpeg for gif/mp4 in webui
apt-get install wget ffmpeg -yq
# mount for persistence
mkdir -vp /data/Deforum
test -h ! /data/Deforum && ln -sT /data/Deforum "${ROOT}/models/Deforum"
echo "< Deforum dependencies"
```

**Comment by DevilaN** [comment]
@timelf123:
apt-get is really working? Weren't you running `apt-get update` by hand so there is apt cache leftover in container?
In a few weeks apt cache will change and new version of packages will show up. Then you will not be able to install new packages and `apt-get update` will be necessary. This is mostly the reason for having all of dependencies installed in one step in Dockerfile, but here you are unable to tinker with image layers as this is only a script running inside a container.
Also `apt-get update` later could have impact on currently installed packages (i.e. newer version of ffmpeg will require other libraries to be upgraded and base installation is depending on them so you basically are "playing with fire" a bit).

It is probably working well right now. I'd like to let you know that you might have problems later and why.
